nexus/src/db/datastore.rs

```diff
@@ -238,6 +242,173 @@ impl DataStore {
+        // interaction between datasets and regions.
+        //
+        // This policy allocates regions to distinct Crucible datasets,
+        // favoring datasets with the smallest existing (summed) region
+        // sizes. Basically, "pick the datasets with the smallest load first".
+        //
+        // Longer-term, we should consider:
+        // - Storage size + remaining free space
+        // - Sled placement of datasets
+        // - What sort of loads we'd like to create (even split across all disks
+        // may not be preferable, especially if maintenance is expected)
+        #[derive(Debug, thiserror::Error)]
+        enum RegionAllocateError {
+            #[error("Not enough datasets for replicated allocation: {0}")]
+            NotEnoughDatasets(usize),
+        }
+        type TxnError = TransactionError<RegionAllocateError>;
+        let params: params::DiskCreate = params.clone();
+        self.pool()
+            .transaction(move |conn| {
+                // First, for idempotency, check if regions are already
+                // allocated to this disk.
+                //
+                // If they are, return those regions and the associated
+                // datasets.
+                let datasets_and_regions = region_dsl::region
+                    .filter(region_dsl::disk_id.eq(disk_id))
+                    .inner_join(
+                        dataset_dsl::dataset
+                            .on(region_dsl::dataset_id.eq(dataset_dsl::id)),
+                    )
+                    .select((Dataset::as_select(), Region::as_select()))
+                    .get_results::<(Dataset, Region)>(conn)?;
+                if !datasets_and_regions.is_empty() {
+                    return Ok(datasets_and_regions);
+                }
+
+                let datasets: Vec<Dataset> = dataset_dsl::dataset
+                    // We look for valid datasets (non-deleted crucible datasets).
+                    .filter(dataset_dsl::time_deleted.is_null())
+                    .filter(dataset_dsl::kind.eq(DatasetKind(
+                        crate::internal_api::params::DatasetKind::Crucible,
+                    )))
+                    // Next, observe all the regions allocated to each dataset, and
+                    // determine how much space they're using.
+                    //
+                    // TODO: We could store "free/allocated" space per-dataset,
+                    // and keep them up-to-date, rather than trying to recompute
+                    // this.
+                    //
+                    // TODO: We admittedly don't actually *fail* any request for
+                    // running out of space - we try to send the request down to
+                    // crucible agents, and expect them to fail on our behalf in
+                    // out-of-storage conditions. This should undoubtedly be
+                    // handled more explicitly.
+                    .left_outer_join(
```

Comment:
What's the index behavior for this query?

To make it more concrete: let's pick a realistic number of datasets and regions for a full rack.  How many rows will this query scan?  How does that scale as number of datasets and regions goes up?

Reply:
I went ahead and optimized this to avoid a FULL SCAN, by storing a `size_used` column to the Dataset table which is now being maintained as an index.